File: homework/hw2/c++/queens_genetic.cpp
#include <ctime>
#include <random>
#include <cmath>
#include <algorithm>
#include <iostream>

#include "interface/population.hpp"
#include "utils/random_variables.hpp"

// 将n皇后问题备选解集合建模为种群，基因序列类型为int数组
class QueensPopulation : public PopulationBase<std::vector<int> >{
private:

    using ChromosomeType = std::vector<int>;
    
    // 当前种群中的所有个体
    std::vector<ChromosomeType> _population;

    // 每个个体的适应度
    std::vector<double> _adaptability;

    // 突变概率
    double mutation_rate;

    // 当前种群中最优的个体和次优的个体序号
    int best_index, second_index;

    // 评估一个基因序列的适应度
    double evaluate_chromosome(const ChromosomeType& c) const {
        int n_queens = c.size();
        std::vector<int> lr(n_queens << 1, 0), rl(n_queens << 1, 0);
        std::vector<int> column(n_queens, 0);
        int conflicts = 0;
        for (int i = 0; i < n_queens; ++ i){
            ++ column[c[i]];
            ++ rl[c[i]+i];
            ++ lr[c[i]-i+n_queens-1];   
        }
        for (int i = 0; i < n_queens; ++ i){
            conflicts += column[c[i]] + rl[c[i]+i] + lr[c[i]-i+n_queens-1] - 3;
        }
        // 目标适应度：0
        return -conflicts;
    }
    
    void update_adaptability(){
        for (int i = 0; i < _population.size(); ++ i){
            _adaptability[i] = evaluate_chromosome(_population[i]);
        }
    }

    // 选择亲本，选择适应度最高的两个个体作为下一代亲本
    const ChromosomeType& select_chromosome_with_adaptability() const {
        
        int coin = RandomVariables::uniform_int() & 1;
        return coin == 0 ? _population[second_index] : _population[best_index];
    }

    // 亲本交叉操作，随机选一个点断开，交叉互换
    ChromosomeType cross(const ChromosomeType& c1, const ChromosomeType& c2) const {
        int index = RandomVariables::uniform_int() % c1.size();
        ChromosomeType result(c1.begin(), c1.begin() + index);
        result.insert(result.end(), c2.begin() + index, c2.end());
        return result;
    }

    // 变异
    void mutate(ChromosomeType& c) const {
        // 随机变异一行皇后的位置
        int index = RandomVariables::uniform_int() % c.size();
        c[index] = RandomVariables::uniform_int() % c.size();
    }

    void update_best_and_second(){
        for (int i = 0; i < _adaptability.size(); ++ i){
            if (_adaptability[i] > _adaptability[best_index]){
                second_index = best_index;
                best_index = i;
            } else if (_adaptability[i] > _adaptability[second_index]){
                second_index = i;
            }
        }
    }

public:

    inline std::vector<ChromosomeType> population() const override {return _population;}
    inline std::vector<double> adaptability() const override {return _adaptability;}
    
    // 通过选择亲本以及交叉生成下一代种群
    void cross() override {
        update_best_and_second();
        std::vector<ChromosomeType> new_population(_population.size());
        ChromosomeType c1, c2;
        for (int i = 0; i < _population.size(); ++ i){
            c1 = select_chromosome_with_adaptability();
            c2 = select_chromosome_with_adaptability();
            new_population[i] = cross(c1, c2);
        }
        _population = new_population;
        update_adaptability();
    }
    
    // 对种群中成员进行变异操作
    void mutate() override {
        ChromosomeType new_c;
        for (int i = 0; i < _population.size(); ++ i){
            if (RandomVariables::uniform_real() > mutation_rate){
                continue;
            }
            new_c = _population[i];
            mutate(new_c);
            _population[i] = new_c;
        }
        update_adaptability();
    }
    
    void show() const override {
        for (int i = 0; i < _population.size(); ++ i){
            std::cout << "<" << i << "> Adaptability: " << _adaptability[i] << std::endl;
            
            // 只输出解
            if (_adaptability[i] == 0){
                std::cout << "Queens: " ;
                for (auto x : _population[i]){
                    std::cout << x << ' ';
                }
                std::cout << std::endl;
            }
        }
    }

    QueensPopulation(const std::vector<ChromosomeType>& p, double _mutation_rate) : _population(p), 
        mutation_rate(_mutation_rate), _adaptability(p.size()), best_index(0), second_index(0) {
        update_adaptability();
    }
};

// 生成初始种群
std::vector<std::vector<int> > generate_queens_population(int n_queens, int population_size){
    std::vector<std::vector<int> > p(population_size);

    for (int i = 0; i < population_size; ++ i){
        p[i] = RandomVariables::uniform_permutation(n_queens);
    }
    return p;
}

#include "algorithm/genetic.hpp"

int n = 20;

int main(){

    time_t t0 = time(nullptr);

    std::ios::sync_with_stdio(false);

    // n皇后，种群大小8n
    auto p = generate_queens_population(n, n << 3);
    // 突变概率0.9
    QueensPopulation pop(p, 0.9);
    GeneticAlgorithm<QueensPopulation> gen(pop);
    // 进化8n代
    gen.evolve(n << 3);
    
    std::cout << "Total time: " << time(nullptr) - t0 << std::endl;
    
    return 0;
}

File: homework/hw2/python/queens_genetic.py
from time import time
from typing import List

from interface.population import PopulationBase
from utils.random_variables import RandomVariables
from algorithm.genetic import GeneticAlgorithm


class QueensPopulation(PopulationBase):
    ChromosomeType = List[int]

    def __init__(self, p: List[ChromosomeType], mutation_rate: float):
        self._population = p
        self._mutation_rate = mutation_rate
        self._adaptability = [0] * len(p)
        self._update_adaptability()
        self._best_index, self._second_index = 0, 0
        self.end_signal = False

    def _evaluate_chromosome(self, c: ChromosomeType) -> float:  # c：n个数字的数组 作为染色体

        column = [0] * len(c)
        lr = [0] * (len(c) << 1)
        rl = [0] * (len(c) << 1)

        conflicts = 0

        for i in range(len(c)):
            column[c[i]] += 1
            rl[c[i] + i] += 1
            lr[c[i] - i + len(c) - 1] += 1

        conflicts = sum(column[c[i]] + rl[c[i] + i] + lr[c[i] - i + len(c) - 1] - 3 for i in range(len(c)))
        if conflicts == 0:
            self.end_signal = True
        return -conflicts  # 将负的conflicts作为adaptability

    def _update_adaptability(self) -> None:
        self._adaptability = [self._evaluate_chromosome(c) for c in self._population]

    def _select_chromosome_with_adaptability(self) -> ChromosomeType:
        coin = RandomVariables.uniform_int() & 1
        return self._population[self._second_index] if coin == 0 else self._population[self._best_index]

    @staticmethod
    def _cross(c1: ChromosomeType, c2: ChromosomeType) -> ChromosomeType:  # 选择第一个个体随机前index个基因与第二个个体后n-index个基因杂交
        index = RandomVariables.uniform_int() % len(c1)
        result = c1[:index] + c2[index:]
        return result

    @staticmethod
    def _mutate(c: ChromosomeType) -> None:  # 随机index随机值
        index = RandomVariables.uniform_int() % len(c)
        c[index] = RandomVariables.uniform_int() % len(c)

    def _update_best_and_second(self) -> None:  # 正如其名，按adaptability 更新self._best_index 和second
        for i in range(len(self._adaptability)):
            if self._adaptability[i] > self._adaptability[self._best_index]:
                self._second_index = self._best_index
                self._best_index = i
            elif self._adaptability[i] > self._adaptability[self._second_index]:
                self._second_index = i

    def population(self) -> List[ChromosomeType]:
        return self._population

    def adaptability(self) -> List[float]:
        return self._adaptability

    def cross(self) -> None:
        self._update_best_and_second()

        self._population = [
            self._cross(self._select_chromosome_with_adaptability(), self._select_chromosome_with_adaptability())
            for i in range(len(self._population))
        ]  # population总数不变，不过都是最好的两个的儿子。

        self._update_adaptability()

    def mutate(self) -> None:

        for i in range(len(self._population)):
            if RandomVariables.uniform_real() > self._mutation_rate:
                continue

            self._mutate(self._population[i])

        self._update_adaptability()

    def show(self) -> None:
        for i in range(len(self._population)):
            print(f"<{i}> Adaptability: {self._adaptability[i]}")

            if self._adaptability[i] == 0:
                print(f"Queens: {self._population[i]}")


if __name__ == "__main__":
    t0 = time()

    n = 63
    p = [RandomVariables.uniform_permutation(n) for i in range(n << 3)]  # 种群数量n*8
    pop = QueensPopulation(p, 0.9)  # 有0.9的突变率

    gen = GeneticAlgorithm(pop)

    gen.evolve(n << 3)  # 种群演化总代数n*3

    print(f"Total time: {time() - t0}")
